Optimize model/nodes.js with string functions and other tricks

zhujinxuan opened the pull request:
## Problems:

## Slow Immutable or Array operations; Use string as a solution in recursive call
Here I add getPathAsString.  Because string operations (at least in v8 using ConString to join strings) are much faster than array.push and List.push.  Then use a string in recursive increment is faster than getAncestors and getPath directly.  Here I add getPathAsString and switch many code to path approach.

This PR switch to a string method as `${num1} ${num2}` to store path.

## toArray is slow
Most data are store as immutable type, like marks(set), characters (list), nodes(list). Using `toArray` functions has to convert from immutable tree structure to arrays again and again.  These toArray functions slow down the performance.
Update: toArray is slow, but it is slower to create a new List object.  I think we shall create and cache List object when necessary.  Thus `asArray` methods are still kept.

## searching in getKeys is as slow as DFS
search in getKeys is O(n = number of descendants).  Because keys is neither hashed nor ordered for quick lookup, these search are as bad as DFS.  This PR removes getKeys and getKeysAsArray

## getCharacters(AsArray) is perhaps un-necessary to be cached
Each character is small. And caching arrays or lists based on a lot of characters can perhaps consume too much memory.  I suggest we can cache `getTexts[AtRange][AsArray]`. But it is perhaps not good to cache getCharacters

## getCharacters are slow
Here we rewrite get[Active]Marks[AtRange] to avoid call getCharacters

## cache of get[Something]AtRange is not always working well
Even with same [edge]Key and [edge]Offset, range can have different.  We use the `get[Something][At|Between]Position[s]` as proxy to cache results.

# Additional Small Fixes:
1. `getClosest()`
`.rest()` creates a new `List` (O(log(n) + class creation), which is a bit slow
2. getNext/PreviousText
use index instead of generating new list
3. use one concat for getTextsAsArray()

# Benchmark:
See https://github.com/ianstormtaylor/slate/pull/1771#issuecomment-380612708

# Unfixed:
1. node.getLeaves are not cached:
```
    const decs = decorations.filter(d => {
      const { startKey, endKey } = d
      if (startKey == key || endKey == key) return true
      const startsBefore = document.areDescendantsSorted(startKey, key)
      if (!startsBefore) return false
      const endsAfter = document.areDescendantsSorted(key, endKey)
      return endsAfter
    })

    const leaves = node.getLeaves(decs)
```
each time desc are with different references... #

zhujinxuan commented:
1. [Done] toArray functions are removed

zhujinxuan commented:
2. [Done] getKeys are removed~

zhujinxuan commented:
Benchmark will use the benchmark of https://github.com/ianstormtaylor/slate/pull/1765

Because the original benchmark does not reset Memorization, as before and after functions are run before and after suite, not around bench

zhujinxuan commented:
Hypothesis: perhaps withMutation+push(or concat in immutable v4) is faster than asArray + push/concat.  But I would think about it in another PR

zhujinxuan commented:
Based on on-working benchmark in #1765 
Here is the benchmark result:

```
yarn run v1.5.1
$ mkdir -p ./tmp && node ./node_modules/.bin/_matcha --reporter ./support/benchmark/reporter ./tmp/benchmark/*/index.js > ./tmp/benchmark-comparison.json && cross-env BABEL_ENV=test babel-node ./support/benchmark/compare

  benchmarks
    html-serializer
      deserialize
        10.33 → 12.28 ops/sec
      serialize
        1030.88 → 1054.02 ops/sec
    plain-serializer
      deserialize
        543.35 → 634.76 ops/sec
      serialize
        8754.10 → 10853.34 ops/sec
    rendering
      normal
        79.90 → 115.11 ops/sec (44% faster)
    changes
      delete-backward
        12383.80 → 14262.35 ops/sec
      delete-forward
        3178.25 → 4871.16 ops/sec (53% faster)
      insert-text-by-key-multiple
        110.17 → 119.71 ops/sec
      insert-text-by-key
        762.08 → 920.86 ops/sec
      insert-text
        1798.25 → 1400.13 ops/sec
      normalize
        1293.23 → 1128.18 ops/sec
      split-block
        70.53 → 80.05 ops/sec
    models
      from-json-big
        4.91 → 4.73 ops/sec
      from-json
        261.71 → 291.63 ops/sec
      get-blocks-at-range
        1364.93 → 2028.39 ops/sec (49% faster)
      get-blocks
        10448.99 → 9886.83 ops/sec
      get-characters-at-range
        336.25 → 2437.54 ops/sec (625% faster) 🙌
      get-characters
        4784.86 → 3685.55 ops/sec
      get-inlines-at-range
        671.04 → 786.28 ops/sec
      get-inlines
        14650.50 → 15894.19 ops/sec
      get-leaves
        21575.23 → 30745.38 ops/sec (43% faster)
      get-marks-at-range
        277.11 → 846.54 ops/sec (205% faster) 🙌
      get-marks
        890.79 → 990.18 ops/sec
      get-path
        2116.89 → 1856.98 ops/sec
      get-texts-at-range
        4158.91 → 6191.14 ops/sec (49% faster)
      get-texts
        11347.94 → 17578.42 ops/sec (55% faster)
      has-node-multiple
        1410.13 → 2315.14 ops/sec (64% faster)
      has-node
        8035.35 → 12246.88 ops/sec (52% faster)
      to-json
        915.35 → 1657.06 ops/sec (81% faster)
      update-node
        4626.77 → 5135.28 ops/sec

✨  Done in 70.92s.
```

Soreine commented on packages/slate/src/models/node.js:
```diff
@@ -198,11 +197,20 @@ class Node {
+    }
 
+    const firstIndex = this.nodes.indexOf(firstAncestor)
+    const secondIndex = this.nodes.indexOf(secondAncestor)
```
Maybe that's something we should come back to, after we worked through `getAncestors` and the likes. Maybe #1763 will do something about it ?

zhujinxuan replied:
I have put the `first === second` prior to the rest in the recent commit.  But there is something I am not sure:

Because slate memoize getAncestors, the recursive descendant way is almost as fast as the getCommonAncestor way.  However, if we use the recursive descendant method, all common ancestors will memoize the areSorted result.  

I think it perhaps depends on which one we want to cache, areSorted of all common ancestors., or getCommonAncestor

Soreine replied:
Well, if `getCommonAncestor` is rewritten to use `getAncestors` (see [this PR](https://github.com/ianstormtaylor/slate/pull/1763/files#diff-54c7838b3ea52077be8fa46cee479d9bR596)), and we only cache `getAncestors`, then the choice is between caching `areSorted` on common ancestors, or caching `getAncestors` of both descendant. Caching `getAncestors` seems better, especially considering that the input space of `areSorted` is very large and less likely to yield cache hits (3 arguments to cache: `this`, `first`, `second`).

Soreine replied:
OK I re-read your comment. `areSorted` does not seem worthy to memoize. I don't think we need to memoize `getCommonAncestor` either. Both methods have low complexity, since they scale with document depth, which is often shallow. Am I right ?

zhujinxuan replied:
My mistake.  I thought getFurthestAncestor calls getAncestors, which does not indeed.  I will use different ancestor version.

zhujinxuan replied:
I am confused about getCommonAncestors; why not use getAncestors in getCommonAncestors?

zhujinxuan commented on .tern-port:
```diff
@@ -0,0 +1 @@
+55257
```
No problem.

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -373,23 +381,16 @@ class Node {
-    }, [])
+    const empty = new List()
+    const { nodes } = this
+    return empty.withMutations(result => {
```
I will measure the speed up of `get*` a moment later in this PR comments (based on the benchmark I finished yesterday).

I will also make a toy script just to compare the efficiency between toArray and withMutations.

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -400,31 +401,19 @@ class Node {
   getBlocksAtRange(range) {
-    const array = this.getBlocksAtRangeAsArray(range)
-    // Eliminate duplicates by converting to an `OrderedSet` first.
-    return new List(new OrderedSet(array))
```
Because `getBlocksAsArray` depends on `getBlocksAsArray`.  But I delete the `getBlocksAsArray` to avoid `toArray` method.  So I think I shall delete `getBlocksAtRangeAsArray` as well.

zhujinxuan replied:
Perhaps the old `getBlocksAtRange` can duplicated code?  Let me check about whether the new getBlocksAtRange will duplicate blocks.

zhujinxuan replied:
I cannot understand why we need this in the old getBlocks either.  I think perhaps it is caused by copying code from getMarksAtRangeAsArray.

Soreine replied:
Yes it probably is. Good we removed it.

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -400,31 +401,19 @@ class Node {
 
     // PERF: the most common case is when the range is in a single block node,
     // where we can avoid a lot of iterating of the tree.
-    if (startKey == endKey) return [startBlock]
```
Sorry, I did not get `This check was good here (just return a List).`.  I think I return the `List.of(startBlock)`

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -1061,120 +1009,112 @@ class Node {
+    if (!first || !first.marks) return new Set()
 
-    return char.marks.toArray()
+    const empty = new Set()
```
No problem. I will fix other `new Immutable` as well.

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -1061,120 +1009,112 @@ class Node {
+    const empty = new Set()
+
+    return first.marks.withMutations(result => {
+      chars.slice(1).find(char => {
```
I think we can even do not skip the first one, just remove `slice`
I think Immutable-JS has optimized `x.intersect(x)`

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -1061,120 +1009,112 @@ class Node {
 
   /**
-   * Get a set of marks in a `range`, by treating it as collapsed.
+   * Get a set of marks in a `position`, by treating it as collapsed.
```
Great.  Thank you. Fixed~

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -1061,120 +1009,112 @@ class Node {
+      if (this.getClosestBlock(key) !== this.getClosestBlock(previous.key)) {
+        return new OrderedSet()
       }
       const char = previous.characters.get(previous.text.length - 1)
```
Of course. Fixed~

zhujinxuan commented on packages/slate/src/models/text.js:
```diff
@@ -519,14 +533,12 @@ Text.prototype[MODEL_TYPES.TEXT] = true
   'getDecorations',
   'getLeaves',
+  'getMarks',
+  'getMarksAsArray',
```
Fixed~

zhujinxuan commented on packages/slate/src/models/node.js:
```diff
@@ -1704,15 +1659,16 @@ class Node {
+        if (this.key === node.key || this.hasDescendant(desc.key)) {
+          return desc.regenerateKey()
+        }
+        return desc
```
Great find.  Fixed~

zhujinxuan commented:
@Soreine 
get-blocks are speed up by `withMutations`; other methods are slow down with `withMutations`

```
  benchmarks
    changes
    models
      get-blocks-at-range
        user : 659.85 → 964.44 ops/sec (32% faster)
        hr : 1270.18 → 1657.46 ops/sec
      get-blocks
        user : 8161.43 → 7819.18 ops/sec
        hr : 11678.36 → 10760.14 ops/sec
      get-characters-at-range
        user : 259.51 → 207.40 ops/sec
        hr : 321.67 → 304.65 ops/sec
      get-characters
        user : 4519.29 → 3831.11 ops/sec
        hr : 5688.04 → 4819.94 ops/sec
      get-inlines-at-range
        user : 595.33 → 181.04 ops/sec (229% slower) 😟
        hr : 815.52 → 248.49 ops/sec (228% slower) 😟
      get-inlines
        user : 12584.28 → 2218.70 ops/sec (467% slower) 😟
        hr : 16918.07 → 3512.27 ops/sec (382% slower) 😟
      get-leaves
        user : 9772.52 → 10533.04 ops/sec
        hr : 16600.33 → 18558.54 ops/sec
      get-marks-at-range
        user : 185.23 → 178.57 ops/sec
        hr : 263.01 → 255.47 ops/sec
      get-marks
        user : 738.12 → 499.08 ops/sec (48% slower)
        hr : 1040.45 → 783.30 ops/sec
      get-path
        user : 2356.79 → 2580.05 ops/sec
        hr : 3185.16 → 3429.66 ops/sec
      get-texts-at-range
        user : 4450.27 → 237.78 ops/sec (1772% slower) 😱
        hr : 6141.74 → 359.09 ops/sec (1610% slower) 😱
      get-texts
        user : 12558.34 → 12731.59 ops/sec
        hr : 17294.24 → 17381.71 ops/sec
```

zhujinxuan commented:
Record of calculation:
1. getBlocks: withMutations is faster than asArray.  flatMap method not tried
2. getTexts: asArray is faster than withMutation and flatMap method

zhujinxuan commented:
restore to 47634b8; do not remove asarray metods other than getBlocks
```

  benchmarks
    changes
    models
      get-blocks-at-range
        user : 659.85 → 1019.95 ops/sec (35% faster)
        hr : 1270.18 → 1572.72 ops/sec
      get-blocks
        user : 8161.43 → 7510.32 ops/sec
        hr : 11678.36 → 9395.82 ops/sec
      get-characters-at-range
        user : 259.51 → 2122.89 ops/sec (88% faster)
        hr : 321.67 → 2797.66 ops/sec (89% faster)
      get-characters
        user : 4519.29 → 3945.76 ops/sec
        hr : 5688.04 → 4893.22 ops/sec
      get-inlines-at-range
        user : 595.33 → 718.23 ops/sec
        hr : 815.52 → 917.85 ops/sec
      get-inlines
        user : 12584.28 → 12479.70 ops/sec
        hr : 16918.07 → 15280.66 ops/sec
      get-leaves
        user : 9772.52 → 9285.54 ops/sec
        hr : 16600.33 → 15358.93 ops/sec
      get-marks-at-range
        user : 185.23 → 389.18 ops/sec (52% faster)
        hr : 263.01 → 673.36 ops/sec (61% faster)
      get-marks
        user : 738.12 → 618.38 ops/sec
        hr : 1040.45 → 873.07 ops/sec
      get-path
        user : 2356.79 → 2193.39 ops/sec
        hr : 3185.16 → 2874.73 ops/sec
      get-texts-at-range
        user : 4450.27 → 4355.67 ops/sec
        hr : 6141.74 → 5426.04 ops/sec
      get-texts
        user : 12558.34 → 11839.11 ops/sec
        hr : 17294.24 → 13827.12 ops/sec
```

zhujinxuan commented:
```
  benchmarks
    changes
    models
      get-blocks-at-range
        user : 659.85 → 1076.28 ops/sec (39% faster)
        hr : 1270.18 → 1785.42 ops/sec (29% faster)
      get-blocks
        user : 8161.43 → 7308.74 ops/sec
        hr : 11678.36 → 9702.35 ops/sec
      get-characters-at-range
        user : 259.51 → 1554.01 ops/sec (83% faster)
        hr : 321.67 → 2294.42 ops/sec (86% faster)
      get-characters
        user : 4519.29 → 3862.09 ops/sec
        hr : 5688.04 → 4788.58 ops/sec
      get-inlines-at-range
        user : 595.33 → 854.42 ops/sec (30% faster)
        hr : 815.52 → 1054.79 ops/sec
      get-inlines
        user : 12584.28 → 13198.36 ops/sec
        hr : 16918.07 → 19213.09 ops/sec
      get-leaves
        user : 9772.52 → 10468.13 ops/sec
        hr : 16600.33 → 19557.47 ops/sec
      get-marks-at-range
        user : 185.23 → 368.13 ops/sec (50% faster)
        hr : 263.01 → 685.06 ops/sec (62% faster)
      get-marks
        user : 738.12 → 727.68 ops/sec
        hr : 1040.45 → 1099.90 ops/sec
      get-path
        user : 2356.79 → 2355.85 ops/sec
        hr : 3185.16 → 3380.45 ops/sec
      get-texts-at-range
        user : 4450.27 → 3975.01 ops/sec
        hr : 6141.74 → 5945.50 ops/sec
      get-texts
        user : 12558.34 → 13240.48 ops/sec
        hr : 17294.24 → 18263.71 ops/sec
```

faster getInlineAtRange

zhujinxuan commented:
Hypo: getKeysAsSet speeds up 
without getKeysAsSet
```
benchmarks
    changes
      delete-backward
        user : 572.69 → 517.81 ops/sec
        hr : 872.94 → 926.06 ops/sec
      delete-forward
        user : 111.47 → 123.04 ops/sec
        hr : 189.06 → 199.47 ops/sec
      insert-text-by-key-multiple
        user : 79.29 → 79.52 ops/sec
        hr : 112.38 → 117.00 ops/sec
      insert-text-by-key
        user : 619.38 → 628.92 ops/sec
        hr : 909.81 → 903.22 ops/sec
      insert-text
        user : 1839.59 → 1916.69 ops/sec
        hr : 3002.62 → 3109.28 ops/sec
      normalize
        user : 776.55 → 749.30 ops/sec
        hr : 1119.20 → 1127.26 ops/sec
      split-block
        user : 482.76 → 501.67 ops/sec
        hr : 735.60 → 773.29 ops/sec
    models
      get-blocks-at-range
        user : 605.23 → 1466.36 ops/sec (59% faster)
        hr : 1130.79 → 2018.13 ops/sec (44% faster)
      get-blocks
        user : 6816.23 → 7383.16 ops/sec
        hr : 10862.01 → 10023.22 ops/sec
      get-characters-at-range
        user : 261.97 → 1920.03 ops/sec (86% faster)
        hr : 324.43 → 2349.67 ops/sec (86% faster)
      get-characters
        user : 4323.56 → 3497.34 ops/sec
        hr : 5337.70 → 4542.04 ops/sec
      get-inlines-at-range
        user : 626.49 → 758.37 ops/sec
        hr : 820.07 → 951.14 ops/sec
      get-inlines
        user : 12105.72 → 11491.82 ops/sec
        hr : 16466.34 → 16720.69 ops/sec
      get-leaves
        user : 9233.74 → 9289.38 ops/sec
        hr : 15257.98 → 16748.95 ops/sec
      get-marks-at-range
        user : 170.35 → 483.34 ops/sec (65% faster)
        hr : 230.68 → 789.09 ops/sec (71% faster)
      get-marks
        user : 488.06 → 721.28 ops/sec (32% faster)
        hr : 609.98 → 1055.75 ops/sec (42% faster)
      get-path
        user : 2183.58 → 2297.97 ops/sec
        hr : 2952.49 → 3126.82 ops/sec
      get-texts-at-range
        user : 3992.27 → 4030.76 ops/sec
        hr : 5585.58 → 5792.61 ops/sec
      get-texts
        user : 11809.77 → 12420.49 ops/sec
        hr : 16383.45 → 17184.29 ops/sec
      has-node-multiple
        user : 2110.64 → 2207.25 ops/sec
        hr : 2397.46 → 2606.46 ops/sec
      has-node
        user : 9601.06 → 9618.93 ops/sec
        hr : 12658.86 → 12747.01 ops/sec
      to-json
        user : 825.00 → 885.15 ops/sec
        hr : 1279.14 → 1366.23 ops/sec
      update-node
        user : 2985.81 → 3162.44 ops/sec
        hr : 4177.48 → 4760.90 ops/sec
```
with getKeysAsSet
```
```

zhujinxuan commented:
Hypo: perhaps we shall we AsSeq rather than AsArray as proxy?

zhujinxuan commented:
I add decoration rendering in benchmark PR; then 098cd47 does not make things faster... strange?

```
  benchmarks
    changes
      delete-backward
        user : 463.24 → 466.96 ops/sec
        hr : 835.62 → 837.20 ops/sec
      delete-forward
        user : 125.42 → 117.19 ops/sec
        hr : 208.84 → 182.87 ops/sec
      insert-text-by-key-multiple
        user : 82.17 → 71.48 ops/sec
        hr : 114.12 → 98.60 ops/sec
      insert-text-by-key
        user : 579.37 → 576.22 ops/sec
        hr : 887.38 → 819.29 ops/sec
      insert-text
        user : 1795.29 → 1701.05 ops/sec
        hr : 3088.39 → 2903.89 ops/sec
      normalize
        user : 684.02 → 695.86 ops/sec
        hr : 1103.73 → 917.88 ops/sec
      split-block
        user : 535.58 → 476.72 ops/sec
        hr : 799.29 → 739.18 ops/sec
    models
      from-json-big
        user : 3.86 → 3.73 ops/sec
        hr : 5.18 → 4.94 ops/sec
      from-json
        user : 178.84 → 167.68 ops/sec
        hr : 268.29 → 265.33 ops/sec
      get-blocks-at-range
        user : 727.40 → 1155.15 ops/sec (37% faster)
        hr : 1033.05 → 1629.77 ops/sec (37% faster)
      get-blocks
        user : 7254.13 → 5905.01 ops/sec
        hr : 7053.70 → 8811.24 ops/sec
      get-characters-at-range
        user : 194.84 → 1694.82 ops/sec (89% faster)
        hr : 171.75 → 2402.93 ops/sec (93% faster)
      get-characters
        user : 3068.16 → 3236.83 ops/sec
        hr : 2747.16 → 4028.15 ops/sec (32% faster)
      get-inlines-at-range
        user : 356.55 → 735.83 ops/sec (52% faster)
        hr : 330.50 → 919.13 ops/sec (64% faster)
      get-inlines
        user : 8568.03 → 10513.91 ops/sec
        hr : 9291.85 → 15966.67 ops/sec (42% faster)
      get-leaves
        user : 8091.79 → 8791.15 ops/sec
        hr : 12704.85 → 17284.79 ops/sec (26% faster)
      get-marks-at-range
        user : 158.44 → 429.01 ops/sec (63% faster)
        hr : 202.17 → 752.93 ops/sec (73% faster)
      get-marks
        user : 556.43 → 765.65 ops/sec (27% faster)
        hr : 792.29 → 1104.52 ops/sec (28% faster)
      get-path
        user : 2052.66 → 2072.71 ops/sec
        hr : 2833.53 → 3065.23 ops/sec
      get-texts-at-range
        user : 3519.21 → 3654.99 ops/sec
        hr : 4978.47 → 5615.73 ops/sec
      get-texts
        user : 8876.56 → 11742.98 ops/sec
        hr : 12141.90 → 16707.73 ops/sec (27% faster)
      has-node-multiple
        user : 1798.09 → 2094.45 ops/sec
        hr : 2046.22 → 2561.02 ops/sec
      has-node
        user : 6871.32 → 8839.55 ops/sec
        hr : 9794.10 → 12750.97 ops/sec
      to-json
        user : 632.11 → 888.78 ops/sec (29% faster)
        hr : 915.15 → 1378.91 ops/sec (34% faster)
      update-node
        user : 2409.56 → 3421.97 ops/sec (30% faster)
        hr : 3322.78 → 4922.54 ops/sec (32% faster)
    html-serializer
      deserialize
        user : 8.44 → 8.00 ops/sec
        hr : 10.42 → 10.07 ops/sec
      serialize
        user : 220.48 → 236.75 ops/sec
        hr : 322.15 → 404.33 ops/sec
    plain-serializer
      deserialize
        user : 388.16 → 407.58 ops/sec
        hr : 574.05 → 597.72 ops/sec
      serialize
        user : 2078.84 → 2303.42 ops/sec
        hr : 2931.31 → 3090.07 ops/sec
    rendering
      normal
        user : 69.69 → 35.55 ops/sec (96% slower)
        hr : 93.46 → 44.60 ops/sec (110% slower) 😟
```

zhujinxuan commented:
TODO:
1. add get-active-marks-at-range;

zhujinxuan commented:
Faster renderText with decorations:
```
  benchmarks
    changes
      delete-backward
        user : 463.24 → 532.80 ops/sec
        hr : 835.62 → 960.16 ops/sec
      delete-forward
        user : 125.42 → 124.01 ops/sec
        hr : 208.84 → 204.40 ops/sec
      insert-text-by-key-multiple
        user : 82.17 → 83.14 ops/sec
        hr : 114.12 → 116.49 ops/sec
      insert-text-by-key
        user : 579.37 → 612.72 ops/sec
        hr : 887.38 → 892.18 ops/sec
      insert-text
        user : 1795.29 → 1768.19 ops/sec
        hr : 3088.39 → 3028.05 ops/sec
      normalize
        user : 684.02 → 747.21 ops/sec
        hr : 1103.73 → 1098.70 ops/sec
      split-block
        user : 535.58 → 535.13 ops/sec
        hr : 799.29 → 802.45 ops/sec
    models
      from-json-big
        user : 3.86 → 4.00 ops/sec
        hr : 5.18 → 5.69 ops/sec
      from-json
        user : 178.84 → 168.06 ops/sec
        hr : 268.29 → 294.58 ops/sec
      get-blocks-at-range
        user : 727.40 → 1234.21 ops/sec (41% faster)
        hr : 1033.05 → 1923.80 ops/sec (46% faster)
      get-blocks
        user : 7254.13 → 6622.62 ops/sec
        hr : 7053.70 → 9512.19 ops/sec (26% faster)
      get-characters-at-range
        user : 194.84 → 1775.49 ops/sec (89% faster)
        hr : 171.75 → 2415.60 ops/sec (93% faster)
      get-characters
        user : 3068.16 → 2764.59 ops/sec
        hr : 2747.16 → 3888.88 ops/sec (29% faster)
      get-inlines-at-range
        user : 356.55 → 690.65 ops/sec (48% faster)
        hr : 330.50 → 886.43 ops/sec (63% faster)
      get-inlines
        user : 8568.03 → 11320.27 ops/sec
        hr : 9291.85 → 16574.67 ops/sec (44% faster)
      get-leaves
        user : 8091.79 → 9510.93 ops/sec
        hr : 12704.85 → 17811.12 ops/sec (29% faster)
      get-marks-at-range
        user : 158.44 → 424.75 ops/sec (63% faster)
        hr : 202.17 → 745.21 ops/sec (73% faster)
      get-marks
        user : 556.43 → 760.97 ops/sec (27% faster)
        hr : 792.29 → 1112.99 ops/sec (29% faster)
      get-path
        user : 2052.66 → 2225.57 ops/sec
        hr : 2833.53 → 3052.88 ops/sec
      get-texts-at-range
        user : 3519.21 → 3690.85 ops/sec
        hr : 4978.47 → 5492.21 ops/sec
      get-texts
        user : 8876.56 → 10686.53 ops/sec
        hr : 12141.90 → 16502.74 ops/sec (26% faster)
      has-node-multiple
        user : 1798.09 → 2100.35 ops/sec
        hr : 2046.22 → 2574.04 ops/sec
      has-node
        user : 6871.32 → 7787.14 ops/sec
        hr : 9794.10 → 12002.06 ops/sec
      to-json
        user : 632.11 → 819.91 ops/sec
        hr : 915.15 → 1348.32 ops/sec (32% faster)
      update-node
        user : 2409.56 → 2800.63 ops/sec
        hr : 3322.78 → 4377.47 ops/sec
    html-serializer
      deserialize
        user : 8.44 → 8.72 ops/sec
        hr : 10.42 → 11.77 ops/sec
      serialize
        user : 220.48 → 277.03 ops/sec
        hr : 322.15 → 464.68 ops/sec (31% faster)
    plain-serializer
      deserialize
        user : 388.16 → 423.83 ops/sec
        hr : 574.05 → 734.13 ops/sec
      serialize
        user : 2078.84 → 2528.09 ops/sec
        hr : 2931.31 → 3898.89 ops/sec
    rendering
      normal
        user : 69.69 → 81.76 ops/sec
        hr : 93.46 → 112.85 ops/sec
```

zhujinxuan commented:
@Soreine I think the further optimization `areDescendantsSorted` comes after your parentMap PR.  Either getDescendant or getAncestors are not fast enough to beat O(n). Perhaps the getAncestors waste too much time in copying lists.  

Therefore, the optimization of `areDescendantsSorted` will be in another PR, if it is still necesssary. (I did a small trick in slate-react to solve the decoration rendering problem)

zhujinxuan commented:
@Soreine @ianstormtaylor I find a crazy fast solution: store path in a string (string operations are much faster than array or immutable-list).  It removes the getAncestors list struction and getPathAsArray's constructing new Array.  

This is just a on working benchmark that I just replaced getPath and getDescendants. Good improvement can be seen in get-path.  Let us see what will happen if we decide to use a path prior in all functions.

```
  benchmarks
    changes
      delete-backward
        user : 375.23 → 451.78 ops/sec
        hr : 724.12 → 797.97 ops/sec
      delete-forward
        user : 125.96 → 132.05 ops/sec
        hr : 202.22 → 225.28 ops/sec
      insert-text-by-key-multiple
        user : 76.16 → 77.31 ops/sec
        hr : 107.04 → 114.15 ops/sec
      insert-text-by-key
        user : 395.60 → 504.15 ops/sec
        hr : 711.92 → 790.28 ops/sec
      insert-text
        user : 1812.50 → 1955.96 ops/sec
        hr : 3295.68 → 3378.41 ops/sec
      normalize
        user : 727.42 → 663.24 ops/sec
        hr : 1208.58 → 1166.33 ops/sec
      split-block
        user : 447.59 → 474.51 ops/sec
        hr : 747.31 → 718.73 ops/sec
    models
      from-json-big
        user : 4.31 → 4.50 ops/sec
        hr : 5.90 → 6.42 ops/sec
      from-json
        user : 173.72 → 174.42 ops/sec
        hr : 281.07 → 304.62 ops/sec
      get-blocks-at-range
        user : 611.28 → 1268.94 ops/sec (52% faster)
        hr : 1206.99 → 1973.09 ops/sec (39% faster)
      get-blocks
        user : 6828.89 → 6474.93 ops/sec
        hr : 8817.76 → 10125.83 ops/sec
      get-characters-at-range
        user : 213.82 → 1748.38 ops/sec (88% faster)
        hr : 268.19 → 2503.71 ops/sec (89% faster)
      get-characters
        user : 3376.64 → 3158.23 ops/sec
        hr : 3992.62 → 4058.24 ops/sec
      get-inlines-at-range
        user : 560.87 → 783.39 ops/sec (28% faster)
        hr : 803.05 → 1006.15 ops/sec
      get-inlines
        user : 11451.04 → 11039.19 ops/sec
        hr : 15943.70 → 16919.53 ops/sec
      get-leaves
        user : 7995.12 → 10693.54 ops/sec (25% faster)
        hr : 14566.82 → 19238.79 ops/sec
      get-marks-at-range
        user : 194.47 → 397.23 ops/sec (51% faster)
        hr : 264.16 → 708.62 ops/sec (63% faster)
      get-marks
        user : 607.89 → 713.18 ops/sec
        hr : 956.18 → 1041.65 ops/sec
      get-path
        user : 1813.17 → 9086.00 ops/sec (80% faster)
        hr : 2755.80 → 12522.73 ops/sec (78% faster)
      get-texts-at-range
        user : 3872.44 → 3730.98 ops/sec
        hr : 5686.90 → 5869.62 ops/sec
      get-texts
        user : 9455.76 → 11598.96 ops/sec
        hr : 14365.11 → 17061.40 ops/sec
      has-node-multiple
        user : 2023.66 → 1965.86 ops/sec
        hr : 2383.93 → 2275.67 ops/sec
      has-node
        user : 7363.82 → 7583.44 ops/sec
        hr : 10587.32 → 11616.72 ops/sec
      to-json
        user : 543.88 → 749.39 ops/sec (27% faster)
        hr : 663.91 → 1340.31 ops/sec (50% faster)
      update-node
        user : 2767.47 → 3314.61 ops/sec
        hr : 3421.65 → 5028.28 ops/sec (32% faster)
    html-serializer
      deserialize
        user : 8.22 → 10.52 ops/sec
        hr : 9.86 → 14.26 ops/sec (31% faster)
      serialize
        user : 229.46 → 294.84 ops/sec
        hr : 377.93 → 497.46 ops/sec
    plain-serializer
      deserialize
        user : 415.20 → 433.48 ops/sec
        hr : 683.41 → 761.94 ops/sec
      serialize
        user : 2419.30 → 2481.57 ops/sec
        hr : 3558.91 → 4032.34 ops/sec
    rendering
      normal
        user : 74.32 → 82.24 ops/sec
        hr : 103.65 → 117.31 ops/sec

✨  Done in 44.01s.
```

zhujinxuan commented:
Current benchmark:
```
  benchmarks
    changes
      delete-backward
        user : 375.23 → 386.13 ops/sec
        hr : 724.12 → 515.03 ops/sec (41% slower)
      delete-forward
        user : 125.96 → 136.05 ops/sec
        hr : 202.22 → 193.61 ops/sec
      insert-text-by-key-multiple
        user : 76.16 → 97.43 ops/sec
        hr : 107.04 → 126.99 ops/sec
      insert-text-by-key
        user : 395.60 → 791.34 ops/sec (50% faster)
        hr : 711.92 → 1093.72 ops/sec (35% faster)
      insert-text
        user : 1812.50 → 2208.53 ops/sec
        hr : 3295.68 → 3353.00 ops/sec
      normalize
        user : 727.42 → 669.03 ops/sec
        hr : 1208.58 → 1050.49 ops/sec
      split-block
        user : 447.59 → 557.15 ops/sec
        hr : 747.31 → 821.63 ops/sec
    models
      from-json-big
        user : 4.31 → 4.20 ops/sec
        hr : 5.90 → 5.67 ops/sec
      from-json
        user : 173.72 → 162.66 ops/sec
        hr : 281.07 → 216.90 ops/sec
      get-blocks-at-range
        user : 611.28 → 1595.62 ops/sec (62% faster)
        hr : 1206.99 → 2227.21 ops/sec (46% faster)
      get-blocks
        user : 6828.89 → 6957.68 ops/sec
        hr : 8817.76 → 10345.00 ops/sec
      get-characters-at-range
        user : 213.82 → 1727.84 ops/sec (88% faster)
        hr : 268.19 → 2359.72 ops/sec (89% faster)
      get-characters
        user : 3376.64 → 3059.29 ops/sec
        hr : 3992.62 → 3984.59 ops/sec
      get-inlines-at-range
        user : 560.87 → 1180.05 ops/sec (52% faster)
        hr : 803.05 → 1539.09 ops/sec (48% faster)
      get-inlines
        user : 11451.04 → 10007.13 ops/sec
        hr : 15943.70 → 16648.63 ops/sec
      get-leaves
        user : 7995.12 → 10007.31 ops/sec
        hr : 14566.82 → 18871.96 ops/sec
      get-marks-at-range
        user : 194.47 → 457.51 ops/sec (57% faster)
        hr : 264.16 → 766.08 ops/sec (66% faster)
      get-marks
        user : 607.89 → 721.08 ops/sec
        hr : 956.18 → 1050.03 ops/sec
      get-path
        user : 1813.17 → 8681.52 ops/sec (79% faster)
        hr : 2755.80 → 12248.09 ops/sec (78% faster)
      get-texts-at-range
        user : 3872.44 → 4156.85 ops/sec
        hr : 5686.90 → 5719.09 ops/sec
      get-texts
        user : 9455.76 → 11566.47 ops/sec
        hr : 14365.11 → 17454.47 ops/sec
      has-node-multiple
        user : 2023.66 → 2081.57 ops/sec
        hr : 2383.93 → 2520.20 ops/sec
      has-node
        user : 7363.82 → 8288.54 ops/sec
        hr : 10587.32 → 11590.09 ops/sec
      to-json
        user : 543.88 → 728.47 ops/sec (25% faster)
        hr : 663.91 → 1264.63 ops/sec (48% faster)
      update-node
        user : 2767.47 → 5033.62 ops/sec (45% faster)
        hr : 3421.65 → 8081.54 ops/sec (58% faster)
    html-serializer
      deserialize
        user : 8.22 → 11.03 ops/sec (25% faster)
        hr : 9.86 → 14.80 ops/sec (33% faster)
      serialize
        user : 229.46 → 272.01 ops/sec
        hr : 377.93 → 461.34 ops/sec
    plain-serializer
      deserialize
        user : 415.20 → 421.08 ops/sec
        hr : 683.41 → 747.78 ops/sec
      serialize
        user : 2419.30 → 2555.43 ops/sec
        hr : 3558.91 → 3929.61 ops/sec
    rendering
      normal
        user : 74.32 → 78.15 ops/sec
        hr : 103.65 → 112.43 ops/sec
```
I think there is still potential to improve.  But I would like a code review for current version.

zhujinxuan commented:
Finished.  It is good for review again.

zhujinxuan commented:
```
  benchmarks
    changes
      delete-backward
        user : 401.05 → 552.30 ops/sec (27% faster)
        hr : 710.27 → 928.70 ops/sec
      delete-forward
        user : 121.85 → 151.17 ops/sec
        hr : 204.48 → 260.93 ops/sec
      insert-text-by-key-multiple
        user : 76.53 → 132.37 ops/sec (42% faster)
        hr : 109.55 → 185.75 ops/sec (41% faster)
      insert-text-by-key
        user : 426.38 → 882.79 ops/sec (52% faster)
        hr : 707.95 → 1389.28 ops/sec (49% faster)
      insert-text
        user : 1905.17 → 2219.20 ops/sec
        hr : 3285.55 → 3980.26 ops/sec
      normalize
        user : 760.10 → 768.18 ops/sec
        hr : 1257.96 → 1186.91 ops/sec
      split-block
        user : 423.77 → 560.83 ops/sec
        hr : 732.10 → 792.81 ops/sec
    models
      from-json-big
        user : 4.49 → 4.52 ops/sec
        hr : 6.37 → 6.35 ops/sec
      from-json
        user : 174.14 → 182.90 ops/sec
        hr : 307.80 → 326.47 ops/sec
      get-blocks-at-range
        user : 672.98 → 1759.14 ops/sec (62% faster)
        hr : 1360.04 → 2978.48 ops/sec (54% faster)
      get-blocks
        user : 7825.45 → 6477.36 ops/sec
        hr : 12028.94 → 10584.92 ops/sec
      get-characters-at-range
        user : 239.72 → 1735.11 ops/sec (86% faster)
        hr : 333.67 → 2446.36 ops/sec (86% faster)
      get-characters
        user : 4484.67 → 3047.04 ops/sec (47% slower)
        hr : 5716.75 → 4028.33 ops/sec (42% slower)
      get-inlines-at-range
        user : 596.54 → 1273.23 ops/sec (53% faster)
        hr : 852.70 → 1563.86 ops/sec (45% faster)
      get-inlines
        user : 9786.15 → 10438.77 ops/sec
        hr : 16352.54 → 16996.80 ops/sec
      get-leaves
        user : 10087.88 → 9074.64 ops/sec
        hr : 17435.30 → 18570.96 ops/sec
      get-marks-at-range
        user : 184.08 → 433.06 ops/sec (57% faster)
        hr : 257.98 → 738.69 ops/sec (65% faster)
      get-marks
        user : 623.86 → 642.84 ops/sec
        hr : 990.68 → 1054.01 ops/sec
      get-path
        user : 2290.81 → 9270.58 ops/sec (75% faster)
        hr : 3330.39 → 12634.78 ops/sec (74% faster)
      get-texts-at-range
        user : 4203.17 → 4237.90 ops/sec
        hr : 6083.41 → 5718.22 ops/sec
      get-texts
        user : 11133.30 → 9571.17 ops/sec
        hr : 16854.39 → 16182.73 ops/sec
      has-node-multiple
        user : 2228.73 → 2106.95 ops/sec
        hr : 2623.18 → 2531.92 ops/sec
      has-node
        user : 8229.98 → 8410.46 ops/sec
        hr : 12397.45 → 12776.77 ops/sec
      to-json
        user : 766.76 → 744.34 ops/sec
        hr : 1240.13 → 1303.73 ops/sec
      update-node
        user : 2963.77 → 5054.36 ops/sec (41% faster)
        hr : 4706.08 → 7935.69 ops/sec (41% faster)
    html-serializer
      deserialize
        user : 9.41 → 10.62 ops/sec
        hr : 12.88 → 14.62 ops/sec
      serialize
        user : 243.25 → 283.80 ops/sec
        hr : 435.67 → 479.81 ops/sec
    plain-serializer
      deserialize
        user : 450.52 → 449.76 ops/sec
        hr : 736.76 → 789.03 ops/sec
      serialize
        user : 2312.58 → 2612.42 ops/sec
        hr : 3686.58 → 4201.58 ops/sec
    rendering
      normal
        user : 76.40 → 77.54 ops/sec
        hr : 109.49 → 113.39 ops/sec

✨  Done in 42.33s.
```
Speed up with getCommonAncestor. Restore the performance of normalize

zhujinxuan commented:
I think we can remove get-characters as public method in node.js and in text.js, mostly because:
1. they are slow, and cost a lot of memory
2. only getActiveMarks depends on this.  But we can easily implement getActiveMark by offset in text.js and remove getCharactersAtRange dependency in node.js
3. If we switch to leaves as data structure of text.js, perhaps it saves a lot of memory (by smaller list), and is faster (by using more string operation than list operation). Removing them can help us restruct the text.js

zhujinxuan commented:
Update: I rewrite the getCommonAncestors to speed up X-at-range changes, and rewrite getActiveMarks to avoid getCharacters.

zhujinxuan commented:
Update: test the bench of `get-active-marks-at-range`

```
  benchmarks
    changes
      delete-backward
        user : 391.90 → 542.49 ops/sec (28% faster)
        hr : 722.63 → 865.94 ops/sec
      delete-forward
        user : 121.52 → 134.18 ops/sec
        hr : 215.87 → 210.32 ops/sec
      insert-text-by-key-multiple
        user : 76.58 → 124.69 ops/sec (39% faster)
        hr : 109.08 → 177.73 ops/sec (39% faster)
      insert-text-by-key
        user : 404.65 → 892.77 ops/sec (55% faster)
        hr : 713.65 → 1323.12 ops/sec (46% faster)
      insert-text
        user : 1998.60 → 2313.89 ops/sec
        hr : 3506.56 → 4015.95 ops/sec
      normalize
        user : 725.80 → 646.32 ops/sec
        hr : 1241.47 → 1170.78 ops/sec
      split-block
        user : 443.50 → 543.25 ops/sec
        hr : 739.62 → 853.78 ops/sec
    models
      from-json-big
        user : 4.45 → 4.45 ops/sec
        hr : 6.32 → 6.29 ops/sec
      from-json
        user : 180.18 → 191.55 ops/sec
        hr : 327.76 → 322.07 ops/sec
      get-active-marks-at-range
        user : 409.70 → 3577.00 ops/sec (89% faster)
        hr : 571.55 → 5685.33 ops/sec (90% faster)
      get-blocks-at-range
        user : 679.03 → 1920.40 ops/sec (65% faster)
        hr : 1315.31 → 3017.87 ops/sec (56% faster)
      get-blocks
        user : 7444.76 → 7517.74 ops/sec
        hr : 12114.91 → 10934.75 ops/sec
      get-characters-at-range
        user : 225.16 → 1863.76 ops/sec (88% faster)
        hr : 321.52 → 2697.60 ops/sec (88% faster)
      get-characters
        user : 3594.19 → 3419.67 ops/sec
        hr : 5177.29 → 4869.78 ops/sec
      get-inlines-at-range
        user : 584.08 → 1139.05 ops/sec (49% faster)
        hr : 834.71 → 1488.36 ops/sec (44% faster)
      get-inlines
        user : 10365.32 → 9913.84 ops/sec
        hr : 16057.17 → 16510.91 ops/sec
      get-leaves
        user : 7851.22 → 11847.71 ops/sec (34% faster)
        hr : 14409.39 → 20733.92 ops/sec (31% faster)
      get-marks-at-range
        user : 200.90 → 548.62 ops/sec (63% faster)
        hr : 268.51 → 864.97 ops/sec (69% faster)
      get-marks
        user : 623.71 → 741.84 ops/sec
        hr : 986.34 → 1076.16 ops/sec
      get-path
        user : 2039.92 → 9258.55 ops/sec (78% faster)
        hr : 3133.64 → 12413.93 ops/sec (75% faster)
      get-texts-at-range
        user : 4380.78 → 4097.21 ops/sec
        hr : 6208.59 → 5694.73 ops/sec
      get-texts
        user : 10258.54 → 11616.70 ops/sec
        hr : 16053.95 → 17446.05 ops/sec
      has-node-multiple
        user : 2081.58 → 2203.88 ops/sec
        hr : 2591.59 → 2556.52 ops/sec
      has-node
        user : 8244.33 → 8423.78 ops/sec
        hr : 12973.01 → 12527.16 ops/sec
      to-json
        user : 731.43 → 824.43 ops/sec
        hr : 1196.07 → 1336.11 ops/sec
      update-node
        user : 3471.47 → 4950.00 ops/sec (30% faster)
        hr : 5095.45 → 7983.88 ops/sec (36% faster)
    html-serializer
      deserialize
        user : 9.50 → 11.08 ops/sec
        hr : 12.93 → 14.96 ops/sec
      serialize
        user : 249.10 → 285.36 ops/sec
        hr : 435.76 → 482.61 ops/sec
    plain-serializer
      deserialize
        user : 434.63 → 450.80 ops/sec
        hr : 781.21 → 788.17 ops/sec
      serialize
        user : 2353.49 → 2593.34 ops/sec
        hr : 3720.90 → 3930.94 ops/sec
    rendering
      normal
        user : 71.85 → 78.68 ops/sec
        hr : 105.75 → 111.24 ops/sec

✨  Done in 43.73s.
```

zhujinxuan commented:
@Soreine Hi, can you add a performance tag in this PR?

zhujinxuan commented:
Current benchmark:
```
  benchmarks
    changes
      delete-backward
        user : 423.69 → 481.18 ops/sec
        hr : 746.78 → 828.24 ops/sec
      delete-forward
        user : 118.91 → 146.09 ops/sec
        hr : 207.58 → 252.72 ops/sec
      insert-text-by-key-multiple
        user : 67.38 → 129.77 ops/sec (48% faster)
        hr : 92.52 → 189.44 ops/sec (51% faster)
      insert-text-by-key
        user : 465.77 → 1239.23 ops/sec (62% faster)
        hr : 729.56 → 1931.50 ops/sec (62% faster)
      insert-text
        user : 1967.27 → 2304.60 ops/sec
        hr : 3328.32 → 3816.94 ops/sec
      normalize
        user : 719.85 → 623.71 ops/sec
        hr : 1182.99 → 1084.10 ops/sec
      split-block
        user : 458.56 → 439.40 ops/sec
        hr : 759.87 → 712.31 ops/sec
    models
      from-json-big
        user : 4.45 → 4.32 ops/sec
        hr : 6.41 → 6.00 ops/sec
      from-json
        user : 177.67 → 184.33 ops/sec
        hr : 315.48 → 265.22 ops/sec
      get-active-marks-at-range
        user : 432.55 → 5791.42 ops/sec (93% faster)
        hr : 568.14 → 8596.37 ops/sec (93% faster)
      get-blocks-at-range
        user : 707.50 → 1997.95 ops/sec (65% faster)
        hr : 1382.50 → 3181.75 ops/sec (57% faster)
      get-blocks
        user : 6458.77 → 6912.38 ops/sec
        hr : 11246.67 → 10295.23 ops/sec
      get-characters-at-range
        user : 244.98 → 2083.08 ops/sec (88% faster)
        hr : 322.23 → 2845.90 ops/sec (89% faster)
      get-characters
        user : 3733.24 → 3725.92 ops/sec
        hr : 5424.73 → 4837.92 ops/sec
      get-inlines-at-range
        user : 613.41 → 1169.20 ops/sec (48% faster)
        hr : 862.22 → 1461.62 ops/sec (41% faster)
      get-inlines
        user : 11341.35 → 10449.58 ops/sec
        hr : 17383.76 → 16409.92 ops/sec
      get-leaves
        user : 8882.30 → 9981.61 ops/sec
        hr : 16612.30 → 18977.80 ops/sec
      get-marks-at-range
        user : 176.01 → 645.79 ops/sec (73% faster)
        hr : 255.44 → 1088.46 ops/sec (77% faster)
      get-marks
        user : 640.24 → 1953.46 ops/sec (67% faster)
        hr : 1015.91 → 2949.35 ops/sec (66% faster)
      get-path
        user : 2108.10 → 18896.33 ops/sec (89% faster)
        hr : 3087.24 → 35343.67 ops/sec (91% faster)
      get-texts-at-range
        user : 3436.70 → 4868.60 ops/sec (29% faster)
        hr : 5570.20 → 6998.10 ops/sec
      get-texts
        user : 11357.19 → 11770.74 ops/sec
        hr : 17086.29 → 16218.59 ops/sec
      has-node-multiple
        user : 2268.24 → 2106.70 ops/sec
        hr : 2715.33 → 2482.88 ops/sec
      has-node
        user : 8457.94 → 25241.69 ops/sec (66% faster)
        hr : 12985.82 → 43841.44 ops/sec (70% faster)
      to-json
        user : 792.31 → 1023.18 ops/sec
        hr : 1261.90 → 1500.10 ops/sec
      update-node
        user : 2954.92 → 4351.43 ops/sec (32% faster)
        hr : 4692.50 → 6724.81 ops/sec (30% faster)
    html-serializer
      deserialize
        user : 9.49 → 10.55 ops/sec
        hr : 12.85 → 14.02 ops/sec
      serialize
        user : 239.11 → 309.66 ops/sec
        hr : 431.54 → 501.32 ops/sec
    plain-serializer
      deserialize
        user : 445.62 → 465.91 ops/sec
        hr : 783.59 → 805.99 ops/sec
      serialize
        user : 2456.18 → 2452.93 ops/sec
        hr : 3825.87 → 3924.25 ops/sec
    rendering
      normal
        user : 75.30 → 79.15 ops/sec
        hr : 107.40 → 105.94 ops/sec

✨  Done in 46.40s.
```

zhujinxuan commented:
Ready to review/merge

zhujinxuan commented:
Ready to review/merge

PS: restore getBlocksAsArray back, because:
```
      get-blocks
        user : 6458.77 → 10006.76 ops/sec (35% faster)
        hr : 11246.67 → 14311.82 ops/sec
```

zhujinxuan commented:
update node improvement by setIn and without getAncestors

```
  benchmarks
    changes
      insert-text-by-key-multiple
        user : 77.25 → 150.22 ops/sec (49% faster)
        hr : 110.01 → 214.56 ops/sec (49% faster)
      insert-text-by-key
        user : 408.53 → 1588.44 ops/sec (74% faster)
        hr : 705.23 → 2465.51 ops/sec (71% faster)
        hr : 717.66 → 619.13 ops/sec
    models
      update-node
        user : 2977.71 → 8124.65 ops/sec (63% faster)
        hr : 4623.50 → 11001.33 ops/sec (58% faster)
✨  Done in 47.79s.
```

ianstormtaylor commented:
@zhujinxuan can you please break this up into smaller pull requests that are more clear. It's hard to review a single huge PR with many, many changes that may or may not increase performance. I'd rather be able to run benchmarks against individual (or similarly grouped) changes to see what is and isn't contributing to performance. Thank you.

ianstormtaylor commented:
Also, please if you're exiting early or writing potentially confusing logic, preface it with `PERF:` like the rest of the codebase does, with an explanation. That way the extra code isn't removed in the future, and people will understand what's going on.

zhujinxuan commented:
@ianstormtaylor Great. I am splitting this PRs and submit different parts in different PRs. (marks: , path: on working, get?AsArray: on working)

zhujinxuan commented:
BTW, I just found I made a mistake in showing the percentage (some 90% percentage shall be 1000%, due to a wrong divider).  The real percentage should be

```
  benchmarks
    changes
      delete-backward
        user : 392.40 → 451.44 ops/sec
        hr : 714.29 → 736.47 ops/sec
      delete-forward
        user : 112.19 → 151.70 ops/sec (35.209425870781104% faster)
        hr : 196.91 → 253.96 ops/sec
      insert-text-by-key-multiple
        user : 68.83 → 145.07 ops/sec (110.78096144444429% faster) 🙌
        hr : 89.18 → 204.44 ops/sec (129.26075080211584% faster) 🙌
      insert-text-by-key
        user : 303.69 → 1802.97 ops/sec (493.6858446744912% faster) 🙌
        hr : 461.06 → 2641.30 ops/sec (472.8702598487443% faster) 🙌
      insert-text
        user : 1633.81 → 2211.76 ops/sec (35.3744869491003% faster)
        hr : 2431.21 → 3782.51 ops/sec (55.581204306808274% faster)
      normalize
        user : 604.54 → 676.91 ops/sec
        hr : 873.72 → 1113.87 ops/sec
      split-block
        user : 382.00 → 504.21 ops/sec
        hr : 520.47 → 735.48 ops/sec (41.30944804347693% faster)
    models
      from-json-big
        user : 4.19 → 4.10 ops/sec
        hr : 5.66 → 5.34 ops/sec
      from-json
        user : 194.22 → 162.17 ops/sec
        hr : 306.30 → 222.42 ops/sec (38% slower)
      get-active-marks-at-range
        user : 277.03 → 4991.14 ops/sec (1701.6658052836854% faster) 😱
        hr : 358.44 → 6475.22 ops/sec (1706.4820491420257% faster) 😱
      get-blocks-at-range
        user : 701.69 → 2635.83 ops/sec (275.6398787936136% faster) 🙌
        hr : 1224.99 → 3521.91 ops/sec (187.5044346219344% faster) 🙌
      get-blocks
        user : 6826.29 → 8398.78 ops/sec
        hr : 11346.45 → 10797.96 ops/sec
      get-characters-at-range
        user : 237.35 → 1657.13 ops/sec (598.1755018133649% faster) 🙌
        hr : 316.32 → 1985.12 ops/sec (527.5703153088062% faster) 🙌
      get-characters
        user : 4160.88 → 2901.61 ops/sec (43% slower)
        hr : 5340.50 → 3607.38 ops/sec (48% slower)
      get-inlines-at-range
        user : 607.91 → 1140.00 ops/sec (87.52775861549802% faster)
        hr : 824.30 → 1448.60 ops/sec (75.7374150877844% faster)
      get-inlines
        user : 10792.36 → 11981.33 ops/sec
        hr : 15819.38 → 16840.44 ops/sec
      get-leaves
        user : 7675.01 → 10117.38 ops/sec
        hr : 13739.84 → 18390.05 ops/sec (33.844663398238986% faster)
      get-marks-at-range
        user : 182.46 → 798.74 ops/sec (337.75680504466635% faster) 🙌
        hr : 236.92 → 1244.50 ops/sec (425.2810907654044% faster) 🙌
      get-marks
        user : 635.54 → 1923.78 ops/sec (202.70067470043506% faster) 🙌
        hr : 937.59 → 2869.79 ops/sec (206.08259645420733% faster) 🙌
      get-path
        user : 2048.76 → 23747.84 ops/sec (1059.1337195799192% faster) 😱
        hr : 2974.48 → 38389.95 ops/sec (1190.6433528483012% faster) 😱
      get-texts-at-range
        user : 4304.12 → 5056.86 ops/sec
        hr : 5756.63 → 7175.73 ops/sec
      get-texts
        user : 11569.49 → 11005.10 ops/sec
        hr : 16824.39 → 15174.28 ops/sec
      has-node-multiple
        user : 1993.28 → 2371.91 ops/sec
        hr : 2339.38 → 2671.53 ops/sec
      has-node
        user : 9483.89 → 21481.95 ops/sec (126.5098528213506% faster) 🙌
        hr : 13070.97 → 39220.29 ops/sec (200.0564744677824% faster) 🙌
      to-json
        user : 663.13 → 811.24 ops/sec
        hr : 1138.64 → 1254.31 ops/sec
      update-node
        user : 2959.45 → 8467.53 ops/sec (186.11791567580195% faster) 🙌
        hr : 4706.97 → 10850.25 ops/sec (130.5143736070599% faster) 🙌
    html-serializer
      deserialize
        user : 9.22 → 10.40 ops/sec
        hr : 12.33 → 13.73 ops/sec
      serialize
        user : 271.35 → 231.91 ops/sec
        hr : 459.95 → 353.93 ops/sec
    plain-serializer
      deserialize
        user : 440.28 → 455.17 ops/sec
        hr : 751.95 → 728.53 ops/sec
      serialize
        user : 2166.33 → 2973.49 ops/sec (37.259700780415464% faster)
        hr : 3468.49 → 4041.76 ops/sec
    rendering
      normal
        user : 75.54 → 75.01 ops/sec
        hr : 106.16 → 105.00 ops/sec

✨  Done in 52.76s.
```

ianstormtaylor commented:
Closing in favor of splitting it up into smaller PRs, thanks @zhujinxuan!
